freshmaker/models.py

```diff
@@ -617,6 +617,33 @@ def get_lowest_build_id(cls, session):
             return 0
         return build.build_id
 
+    @classmethod
+    def get_original_nvr_by_rebuilt_nvr(cls, session, rebuilt_nvr):
+        """
+        Search for an original build event, that has been built by FM, and
+            get original_nvr from it
+
+        :param session: database session
+        :param str rebuilt_nvr: rebuilt NVR to look build by
+        :rtype: str or None
+        :return: original NVR from the first FM build for given NVR
+        """
+        original_nvr = None
+        # build event should be only one in database, or raise an error
+        build_event = db.session.query(ArtifactBuild).filter(
+            ArtifactBuild.rebuilt_nvr == rebuilt_nvr,
+            ArtifactBuild.type == ArtifactType.IMAGE.value,
+        ).distinct().one_or_none()
+        # recursively search for original build event
+        if build_event is not None:
+            original_nvr = build_event.original_nvr
+            next_nvr = cls.get_original_nvr_by_rebuilt_nvr(session,
```

Comment:
Any reason of get the original image recursively? as an example: original or original -> original -> rebuilt_nvr, this will ignore the original image and is not what we want.

Reply:
I am not quite understand. There is always suppose to be previous rebuild event with the **olderEvent.rebuilt_nvr** == **currentlyProcessedEvent.orignial_nvr**. So if there is at least one event with that condition, this method will return **olderEvent.original_nvr**. And the reason of getting it recursively is because it reflects logic how we suppose to search for **original_nvr**, just searching every time previous event untill we can't find any, and return original_nvr of the last found build event.

Reply:
@qixiang the reason is that if Release Driver has rebuilt this image for conflict resolution, the `original_nvr` can actually still be a `rebuilt_nvr`. The recursion will ensure the actual `original_nvr` that is released.

That being said, I noticed a bug in this logic that I had recommended. If `original_nvr` was a previous Freshmaker rebuild that was released, then we actually don't want to continue the recursion. Can you stop the recursion if the event type that triggered the build is `ErrataAdvisoryRPMsSignedEvent` (I think that is the correct name)?

Reply:
No I don't, @mprahl so if I see that event is of type **ErrataAdvisoryRPMsSignedEvent** I should stop and use that event's **original_nvr** or the **original_nvr** of the previously found event?

Reply:
@apaplaus sorry, I meant to say "Could you please stop the recursion....".

That's what I'm thinking. The recursion stops if no build is found or the build was triggered by an event of type `ErrataAdvisoryRPMsSignedEvent`.

Reply:
So will it be enough?
```
        if build_event is not None:
            # check if event that triggered the build is released
            if build_event.event.event_type_id == ErrataAdvisoryRPMsSignedEvent:
                return None
            original_nvr = build_event.original_nvr
            next_nvr = cls.get_original_nvr_by_rebuilt_nvr(session, original_nvr)
```

Reply:
`if build_event.event.event_type_id == ErrataAdvisoryRPMsSignedEvent:` => `if original_nvr and build_event.event.event_type_id == ErrataAdvisoryRPMsSignedEvent:`

Reply:
I didn't get it. I assign **original_nvr** after this conditional so at that point it always will be None. And event if I change their places, why would I check orignal_nvr if I still returning None? You are suggesting to do it like this?:
```
 if build_event is not None:
            original_nvr = build_event.original_nvr
            # check if event that triggered the build is released
            if original_nvr and build_event.event.event_type_id == ErrataAdvisoryRPMsSignedEvent:
                return None
 
            next_nvr = cls.get_original_nvr_by_rebuilt_nvr(session, original_nvr)
```

Reply:
Yes, except that now you can go back to your original proposed if statement of `if build_event.event.event_type_id == ErrataAdvisoryRPMsSignedEvent:`, but return `original_nvr` and not `None`.

Reply:
If we have the following case:

1. `foobar-1-2` is an image rebuilt by product team
2. ` foobar-1-2.xxx` is an image rebuilt for `foobar-1-2`
3. `foobar-1-2.yyy` is an image rebulit for `foobar-1-2.xxx`

I believe when `foobar-1-2.yyy` is shipped, `foobar-1-2.xxx` should always been shipped already, and the bundle image was rebuilt to replace `foobar-1-2` with `foobar-1-2.xxx` already in previous this type of event, so I don't think we need to get the original_nvr recursively.

Reply:
I think we should ask @mprahl. It was his thought originally, so I am not competent enough in this question =)

Reply:
@qixiang the Release Driver case I'm thinking is as follows. Please correct me if I'm wrong.
CVE2 (operator-1.2.124) -> CVE1 (operator-1.2.123)

Release Driver would request a rebuild of operator-1.2.124 to include the CVE fix from operator-1.2.123. CVE2 now has the build of operator-1.2.125 attached. Assume no other changes and the advisories ship.

Once shipped, Freshmaker queries for the original NVR of operator-1.2.125, but the original NVR would be operator-1.2.124 which was never shipped. Freshmaker would then determine no bundle image references the digest of operator-1.2.124 and thus skip any bundle image rebuilds of that. 

Does that make sense?

Reply:
Maybe we should check whether the original_nvr is a published image instead of checking the event type. Because before all this happens, the latest published image can be an image rebuilt by freshmaker, not an image built by product team, so we should stop looking for the original_nvr when an image is a published image.

Even with this, I still see a gap here in your example, both `operator-1-2.123` and `operator-1-2.124` were rebuilt for `operator-1-2`, when `operator-1-2.123` was shipped, the `operator-1-2` in bundle image was replaced by `operator-1-2.123`, and then when `operator-1-2.124` or even `operator-1-2.125` is shipped, the bundle image will not be updated to replace `operator-1-2.123` with `operator-1-2.124` or `operator-1-2.125`, because `operator-1-2.123` is not the original nvr of `operator-1-2.124` or `operator-1-2.125`.

Reply:
@qixiang yes... very good point.

Let's implement your idea of recursively getting the original NVR until the original NVR is published. That is the safest approach.

In addition, lets have Freshmaker look at the advisories that the advisory currently being processed depends on. For all of those advisories, find the container images with the same "package" name and also consider those as operator/operand images to be replaced by the operator/operand image on the advisory being processed.

What do you think?

Reply:
@mprahl I agree, this approach looks good to me.

Reply:
@apaplaus is this approach clear? If not, feel free to reach out over chat.

Reply:
@apaplaus I filed CLOUDWF-4548 to address the second part (traversing the advisory dependency graph). If you can, please implement @qixiang's suggestion of modifying the recursion to stop when there is either no original NVR or an original NVR is found that is published.

Reply:
Gotcha =)